### DISCUSSION TITLE
`vtkCGNSReader` failed to read mesh_NACA0012_inv.su2

### DISCUSSION THREAD

**Discussion by zhang-qiang-github** [su2code/SU2#1598] [Q&A]
**Describe the bug**

I want to use `vtkCGNSReader` to read a `su2` file. But a bug is reported:

```
cgio_file_open: ADF 16: Internal error: Memory boundary tag bad.
```

**To Reproduce**

```
import vtkmodules.all as vtk
reader = vtk.vtkCGNSReader()
reader.SetFileName('mesh_NACA0012_inv.su2`)
reader.Update()
```

**Desktop (please complete the following information):**
 vtk: 9.1.0

**Question**

Is `su2` a standard cgns file?

**Comment by pcarruscag** [comment]
No, it is not.

**Reply by bigfooted** [reply]
if you want to read an SU2 mesh in python or paraview, you can use meshio: https://pypi.org/project/meshio/

**Reply by zhang-qiang-github** [reply]
@bigfooted , Thanks for your reply. But, I find there are some bugs using meshio to read an SU2 mesh.

The example is:

mesh = meshio.read('TestCases\\Tutorials\\compressible_flow\\ActuatorDisk_VariableLoad\\propeller_variable_load.su2')
The reported bug is:

Traceback (most recent call last):
  File "D:\Anaconda\envs\work\lib\site-packages\meshio\_helpers.py", line 103, in _read_file
    return reader_map[file_format](str(path))
  File "D:\Anaconda\envs\work\lib\site-packages\meshio\su2\_su2.py", line 48, in read
    mesh = read_buffer(f)
  File "D:\Anaconda\envs\work\lib\site-packages\meshio\su2\_su2.py", line 102, in read_buffer
    ).reshape(num_verts, dim + extra_columns)
ValueError: cannot reshape array of size 259228 into shape (51846,5)

**Reply by bigfooted** [reply]
this specific su2 file is not in standard su2 format: it has additional information in it that meshio cannot handle (not sure why, I don't know this test case). You have to use a different case or modify the su2 file according to the standard su file format if you want to read it with meshio

**Comment by zhang-qiang-github** [comment]
@pcarruscag If I have a mask of vessel, how can I convert it into `SU2` by vtk?

**Reply by bigfooted** [reply]
I do not know what you mean by 'mask of vessel', or converting it into su2 by vtk.
Can you tell us what your final goal is that you want to achieve?

**Reply by zhang-qiang-github** [reply]
@bigfooted , I cannot upload figure in this reply. Please have a look at the following reply.

**Comment by zhang-qiang-github** [comment]
The mask means a 3D matrix, and the pixel inside the vessel is 1, that outside the vessel is 0. I can extract the surface of vessel, for example:
![image](https://user-images.githubusercontent.com/29271051/162729640-631c6747-3c22-46de-96a8-be9e7fccd3f9.png)


The blood will flow through the vessel, and my final goal is to obtain the pressure under the flowing blood. 

So, I wonder how to convert the 3D mask or surface into `SU2`.

**Reply by bigfooted** [reply]
In what data format is the data now? You could try exporting it to cgns instead, that is a standard that many CAD and meshing software supports.
SU2 can read cgns files directly.

**Reply by zhang-qiang-github** [reply]
I am processing this data in python. The 3D mask matrix is a `numpy.ndarray`, and the extracted surface is `vtkPolyData`.

Actually, the problem is not the data format. The problem is that the `SU2` file should be the finite element grid, and I don't know how to obtain this grid. That's my problem. 

Do you know any python package or any tool to obtain the finite element grid from a `numpy.ndarray` or `vtkPolyData`?

**Reply by bigfooted** [reply]
if you have the point and connectivity information available from the data, you can just write it to a file and that's the entire su2 file format:
https://su2code.github.io/docs/Mesh-File/

**Reply by zhang-qiang-github** [reply]
But, I have the connectivity information for only the surface. I have no information for the point inside the surface. 

So, do you know any tool to generate the connectivity information for the point inside the surface?

**Reply by bigfooted** [reply]
If you want to actually use the internal points, you need a tool that can do a 3D constrained Delaunay triangulation. Most Delaunay triangulations are 2D, due to theoretical limitations of Delaunay in 3D. Paraview has some capabilities, but it needs a lot of work imo to use the surfaces as constraints. you can maybe use cgal, and there are some matlab tools. Frankly, I think it is easier to just discard the internal points and recreate the mesh from the surface.

**Reply by zhang-qiang-github** [reply]
Can `SU2` accept the surface of 3D object to calculate the pressure?

**Reply by bigfooted** [reply]
I don't understand what you want to do. The pressure on the surface is always part of the CFD solution and you can get surface information either directly in SU2 or by a postprocessing tool like paraview.
